cpp/src/Slice/Scanner.l

```diff
@@ -61,144 +39,430 @@ namespace Slice
+    return ICE_STRING_LITERAL;
+}
+
+  /* Matches EOF, but only while scanning a string literal. */
+<STRING_LITERAL><<EOF>> {
+    yy_pop_state();
+    endLocation(yylloc);
+
+    unit->error("encountered EOF while scanning a string literal");
+    return ICE_STRING_LITERAL;
+}
+
+{integer_constant} {
+    setLocation(yylloc);
+
+    IntegerTokPtr itp = new IntegerTok;
+    itp->literal = string(yytext);
+    *yylval = itp;
+    if(!IceUtilInternal::stringToInt64(string(yytext), itp->v))
     {
-        BEGIN(BOMSCAN);
+        assert(itp->v != 0);
+        unit->error("integer constant `" + string(yytext) + "' out of range");
     }
+    return ICE_INTEGER_LITERAL;
 }
 
-"//" {
-    // C++-style comment
-    BEGIN(MAINSCAN);
-    int c;
-    do
+{floating_literal} {
+    setLocation(yylloc);
+
+    errno = 0;
+    FloatingTokPtr ftp = new FloatingTok;
+    *yylval = ftp;
+    string literal(yytext);
+    ftp->literal = literal;
+    char lastChar = literal[literal.size() - 1];
+    if(lastChar == 'f' || lastChar == 'F')
     {
-        c = yyinput();
-        if(c == '\n')
-        {
-            unit->nextLine();
-        }
+        literal = literal.substr(0, literal.size() - 1);    // Clobber trailing 'f' or 'F' suffix
     }
-    while(c != '\n' && c != EOF);
+    ftp->v = strtod(literal.c_str(), 0);
+    if((ftp->v == HUGE_VAL || ftp->v == -HUGE_VAL) && errno == ERANGE)
+    {
+        unit->error("floating-point constant `" + string(yytext) + "' too large (overflow)");
+    }
+    else if(ftp->v == 0 && errno == ERANGE)
+    {
+        unit->error("floating-point constant `" + string(yytext) + "' too small (underflow)");
+    }
+    return ICE_FLOATING_POINT_LITERAL;
+}
+
+  /* ========== Comments ========== */
+
+  /* Matches and consumes a C++ style comment. */
+<*>"//".* {}
+
+  /* Matches the start of a C style comment, and switches the scanner to the C_COMMENT state. */
+<*>"/*" {
+    yy_push_state(C_COMMENT);
+}
+
+  /* Matches any character except for newlines and adds them to the comments. '*' are matched one at a time to ensure
+   * Flex scans '* /' correctly. Flex prioritizes longer matches over shorter ones, so '* /' will match before '*'. */
+<C_COMMENT>"*"    |
+<C_COMMENT>[^\n*] {
+    yymore();
+}
+
+  /* Matches as many newlines as are available and adds them to the comment, after incrementing the line count. */
+<C_COMMENT>\n+ {
+    nextLine(yyleng);
+    yymore();
+}
+
+  /* Matches the end of a C style comment, and reverts the scanner state to what it previously was. */
+<C_COMMENT>"*/" {
+    yy_pop_state();
+
+    string comment(yytext);
+    // The last 2 characters are the '*/' matched by this rule.
+    unit->setComment(comment.substr(0, comment.length() - 2));
+}
+
+  /* Handles reaching EOF while scanning a C style comment by issuing a warning but continuing normally. */
+<C_COMMENT><<EOF>> {
+    yy_pop_state();
+
+    unit->warning(All, "encountered EOF while scanning a comment");
```

Comment:
should this be an error?

Reply:
It was originally a warning, so I left it, just to preserve behavior. I agree it seems more like an error than a warning, but I could see arguments either way.

Reply:
ok let's keep as is for now

Reply:
I changed this to be an error, since that seems to be what we all prefer.